File: tigerline-web/js/harness.js
/* =====================================================================
 * harness.js — risk gate. Never picks bets. Ports tigerline/harness.py.
 * ===================================================================== */
"use strict";

import { dec } from "./decimal.js";

const SKIP_SCENARIOS = new Set(["rotation_trap", "unclear_skip"]);

/**
 * @param {import('./types.js').MatchInput} match
 * @param {{scenario:string, confidence:number, reasons:string[]}} classification
 */
export function evaluateHarness(match, classification) {
  const notes = [];
  const { scenario, confidence } = classification;

  if (SKIP_SCENARIOS.has(scenario)) {
    notes.push(`scenario=${scenario} forces skip`);
    return { adjustment: "skip", notes };
  }

  if (match.risk_flags.includes("favorite_already_qualified")) {
    notes.push("risk flag: favorite already qualified");
    return { adjustment: "skip", notes };
  }

  const favNeed = match.market.favorite === "home" ? match.team_need.home : match.team_need.away;
  const undNeed = match.market.favorite === "home" ? match.team_need.away : match.team_need.home;
  const riskFlagCount = match.risk_flags.length;

  const downgrade = [];
  if (dec.lt(confidence, 0.5)) downgrade.push(`low classifier confidence (${confidence})`);
  if (undNeed === "must_win" && scenario === "two_goal_landing") {
    downgrade.push("underdog must_win against two-goal handicap");
  }
  if (match.group_context.same_time_kickoff && scenario !== "pressure_under") {
    downgrade.push("parallel kickoff in final round");
  }
  if (riskFlagCount >= 2) downgrade.push(`${riskFlagCount} risk flags raised`);

  const upgrade = [];
  if (!downgrade.length) {
    if (dec.gte(confidence, 0.85) && (favNeed === "must_win" || favNeed === "neutral")) {
      upgrade.push("high confidence + favorite has motive");
    }
    if (
      scenario === "two_goal_landing" &&
      dec.gte(confidence, 0.85) &&
      (undNeed === "rotation" || undNeed === "draw_ok")
    ) {
      upgrade.push("clean two-goal landing with passive underdog");
    }
  }

  if (upgrade.length) return { adjustment: "upgrade", notes: [...notes, ...upgrade] };
  if (downgrade.length) return { adjustment: "downgrade", notes: [...notes, ...downgrade] };
  notes.push("no risk signals; proceed at default size");
  return { adjustment: "normal", notes };
}

File: tigerline/harness.py
"""HARNESS-GATE — risk control. Only ever upgrade / downgrade / skip / pass-through.

Never picks the bet. Reads scenario + classifier confidence + team_need + risk
flags and returns a verdict the recommender uses to nudge stake levels.
"""

from __future__ import annotations

from decimal import Decimal

from tigerline.models import (
    HarnessVerdict,
    MatchClassification,
    MatchInput,
)


def evaluate(
    match: MatchInput,
    classification: MatchClassification,
) -> HarnessVerdict:
    """Decide whether to upgrade, normalize, downgrade, or skip the main bet."""
    notes: list[str] = []
    scenario = classification.scenario

    # Hard skip rules — these short-circuit before any upgrade considerations.
    if scenario in ("rotation_trap", "unclear_skip"):
        notes.append(f"scenario={scenario} forces skip")
        return HarnessVerdict(adjustment="skip", notes=notes)

    risk_flag_count = len(match.risk_flags)
    fav_need = match.team_need.home if match.market.favorite == "home" else match.team_need.away
    und_need = match.team_need.away if match.market.favorite == "home" else match.team_need.home

    # Skip when the favorite is already qualified / rotating despite the rule
    # not catching it (e.g. the trap rule's handicap_abs_gte threshold).
    if "favorite_already_qualified" in match.risk_flags:
        notes.append("risk flag: favorite already qualified")
        return HarnessVerdict(adjustment="skip", notes=notes)

    # Downgrade triggers — accumulate, do not short-circuit.
    downgrade_reasons: list[str] = []
    if classification.confidence < Decimal("0.5"):
        downgrade_reasons.append(f"low classifier confidence ({classification.confidence})")
    if und_need == "must_win" and scenario == "two_goal_landing":
        downgrade_reasons.append("underdog must_win against two-goal handicap")
    if match.group_context.same_time_kickoff and scenario != "pressure_under":
        downgrade_reasons.append("parallel kickoff in final round")
    if risk_flag_count >= 2:
        downgrade_reasons.append(f"{risk_flag_count} risk flags raised")

    # Upgrade triggers — only fire when downgrade list is empty.
    upgrade_reasons: list[str] = []
    if not downgrade_reasons:
        if classification.confidence >= Decimal("0.85") and fav_need in ("must_win", "neutral"):
            upgrade_reasons.append("high confidence + favorite has motive")
        if (
            scenario == "two_goal_landing"
            and classification.confidence >= Decimal("0.85")
            and und_need in ("rotation", "draw_ok")
        ):
            upgrade_reasons.append("clean two-goal landing with passive underdog")

    if upgrade_reasons:
        notes.extend(upgrade_reasons)
        return HarnessVerdict(adjustment="upgrade", notes=notes)
    if downgrade_reasons:
        notes.extend(downgrade_reasons)
        return HarnessVerdict(adjustment="downgrade", notes=notes)

    notes.append("no risk signals; proceed at default size")
    return HarnessVerdict(adjustment="normal", notes=notes)


__all__ = ["evaluate"]
